pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java

```diff
@@ -0,0 +1,184 @@
+    final double optimalSegmentSizeBytesMax = desiredSegmentSizeBytes * 1.5;
+
+    if (committingSegmentZKMetadata == null) { // first segment of the partition, hence committing segment is null
+      if (_latestSegmentRowsToSizeRatio > 0) { // new partition group added case
+        long targetSegmentNumRows = (long) (desiredSegmentSizeBytes * _latestSegmentRowsToSizeRatio);
+        targetSegmentNumRows = capNumRowsIfOverflow(targetSegmentNumRows);
+        LOGGER.info(
+            "Committing segment zk metadata is not available, using prev ratio {}, setting rows threshold for {} as {}",
+            _latestSegmentRowsToSizeRatio, newSegmentName, targetSegmentNumRows);
+        return (int) targetSegmentNumRows;
+      } else {
+        final int autotuneInitialRows = streamConfig.getFlushAutotuneInitialRows();
+        LOGGER.info(
+            "Committing segment zk metadata is not available, setting threshold for {} as {}", newSegmentName,
+            autotuneInitialRows);
+        return autotuneInitialRows;
+      }
+    }
+
+    final long committingSegmentSizeBytes = committingSegmentDescriptor.getSegmentSizeBytes();
+    if (committingSegmentSizeBytes <= 0) { // repair segment case
+      final int targetNumRows = committingSegmentZKMetadata.getSizeThresholdToFlushSegment();
+      LOGGER.info(
+          "Committing segment size is not available, setting thresholds from previous segment for {} as {}",
+          newSegmentName, targetNumRows);
+      return targetNumRows;
+    }
+
+    final long timeConsumed = _clock.currentTimeMillis() - committingSegmentZKMetadata.getCreationTime();
+    final long numRowsConsumed = committingSegmentZKMetadata.getTotalDocs();
+    final int numRowsThreshold = committingSegmentZKMetadata.getSizeThresholdToFlushSegment();
+    LOGGER.info(
+        "{}: Data from committing segment: Time {}  numRows {} threshold {} segmentSize(bytes) {}",
+        newSegmentName, TimeUtils.convertMillisToPeriod(timeConsumed), numRowsConsumed, numRowsThreshold,
+        committingSegmentSizeBytes);
+
+    double currentRatio = (double) numRowsConsumed / committingSegmentSizeBytes;
+    // Compute segment size to rows ratio only from the lowest available partition id.
+    // If we consider all partitions then it is likely that we will assign a much higher weight to the most
+    // recent trend in the table (since it is usually true that all partitions of the same table follow more or
+    // less same characteristics at any one point in time).
+    // However, when we start a new table or change controller mastership, we can have any partition completing first.
+    // It is best to learn the ratio as quickly as we can, so we allow any partition to supply the value.
+    int smallestAvailablePartitionGroupId =
+        partitionGroupMetadataList.stream().mapToInt(PartitionGroupMetadata::getPartitionGroupId).min()
+            .orElseGet(() -> 0);
```

Comment:
I know this is just code moved from one place to another but `.orElse(0)` would be better

Reply:
Makes sense, done.